// use-state/dist/index.js
'use strict';

const BasicStore = require('./basic-store');
const StructureStore = require('./structure-store');

var _require = require('./structures/map');

const createMap = _require.createMap;

var _require2 = require('./structures/list');

const createList = _require2.createList;

var _require3 = require('./structures/model');

const createModel = _require3.createModel;

var _require4 = require('./react');

const connect = _require4.connect,
      Provider = _require4.Provider;


function createBasicStore() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  return new (Function.prototype.bind.apply(BasicStore, [null].concat(args)))();
}

function createStructureStore() {
  for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
    args[_key2] = arguments[_key2];
  }

  return new (Function.prototype.bind.apply(StructureStore, [null].concat(args)))();
}

module.exports = {
  Store: StructureStore, createStore: createStructureStore,
  BasicStore: BasicStore, createBasicStore: createBasicStore,
  List: createList, Map: createMap, Model: createModel,
  connect: connect, Provider: Provider
};
